### DISCUSSION TITLE
Some questions about using capnproto in qt

### DISCUSSION THREAD

**Discussion by holdxen** [capnproto/capnproto#2189] [Q&A] [answered]
This is my capnproto file:

![截图 2024-12-01 22-47-09](https://github.com/user-attachments/assets/69620d70-d80f-4e97-a253-5f56ec75b3b4)

This is the code:
![截图 2024-12-01 22-48-49](https://github.com/user-attachments/assets/b7ca8009-54a3-4249-97b7-f1a7884cabba)

*The qt event loop run in main thread, so I start another thread to start my rpc server*

### The first question:
When I got an interface `SomeFunctions` from `Register`, can I save it to the main thread and call the `SomeFunctions::call` function in main thread?
My requirement is that when the user clicks a button, I need to call `SomeFunctions::call`  in qt slot without blocking, that is to say I can't call `Promise::wait` which is blocking. 
So, how I can call the function `SomeFunctions::call`?

### the second question:
Like I said, I start another thread to run my rpc server. I need to notify rpc thread to exit when the qt program exits, but `Promise::wait` is always blocking, how can I do this?

**Comment by kentonv** [comment, marked as answer]
In general, Cap'n Proto RPC objects are only safe to use from the thread where they were created. So, no, you cannot call `SomeFunctions` from another thread.

You can use `kj::Executor` to schedule work on the KJ thread from another thread, or `kj::newPromiseAndCrossThreadFulfiller()` to send a one-way message to the KJ thread. To communicate the opposite direction (invoke the Qt thread from the KJ thread) you will probably have to use Qt APIs.

**Reply by holdxen** [reply]
Thanks for your help
